Fix caml pooled mode and recursive startup

gadmm opened the pull request:
Terminology:
- pooled mode: the option that ensures that all memory allocated with `caml_stat_alloc*` is freed upon shutdown.
- recursive startup: the ability to call `caml_startup*` several times (not technically recursively but whatever) described in the manual.

This PR fixes a crash in the instrumented runtime in pooled mode, and fixes some leaks in the normal runtime in pooled mode or when calling `caml_startup*` several times. It also factors some duplicated startup code and adds some comments to avoid similar issues in the future.

This affects all OCaml versions starting from 4.10.

gadmm committed:
Fix caml pooled mode and recursive startup

caml_init_domain and CAML_EVENTLOG_INIT call caml_stat_alloc
functions, so they are moved after the pool initialisation.

Suubsequently, the call to caml_record_backtrace, which accesses the
Caml_state, must be moved later.

We document that the pool is only available after some point, to hope
avoid similar bugs in the future.

gadmm committed:
Factor startup code

We move duplicated code to startup_aux to simplify further
modifications. This should not change the meaning of the code.

gadmm committed:
Changes

gadmm committed:
Do not reparse OCAMLRUNPARAM on subsequent startups

This commit prevents reparsing OCAMLRUNPARAM on subsequent calls to
caml_startup*. In rare conditions, e.g. when using setenv/putenv to
change the value of caml_use_huge_pages after OCaml has already
started, this can lead to UB (a segfault most probably). This is
unlikely to happen, but it shows that it does not make sense to
reparse OCAMLRUNPARAM. It is unlikely that the reparsing of
OCAMLRUNPARAM was relied on in practice.

gadmm commented:
In addition, the last commit prevents reparsing `OCAMLRUNPARAM` on subsequent calls to `caml_startup*`. In rare conditions, e.g. when using setenv/putenv to change the value of `caml_use_huge_pages` after OCaml has already started, this can lead to UB (a segfault most probably). This is unlikely to happen, but it shows that it does not make sense to reparse `OCAMLRUNPARAM`.

gasche commented on runtime/startup_aux.c:
```diff
@@ -111,8 +116,7 @@ void caml_parse_ocamlrunparam(void)
       case 'a': scanmult (opt, &caml_init_policy); break;
-      case 'b': scanmult (opt, &p); caml_record_backtrace(Val_int (p));
-        break;
+      case 'b': scanmult (opt, &p); record_backtrace = p; break;
```
Given the other cases, this looks like it could be `scanmult (opt, &record_backtrace);`?

gadmm replied:
thanks for catching that!

gasche commented on runtime/startup_byt.c:
```diff
@@ -462,26 +459,11 @@ CAMLexport void caml_main(char_os **argv)
-  caml_gc_message (-1, "### OCaml runtime: debug mode ###\n");
-#endif
-  if (!caml_startup_aux(/* pooling */ caml_cleanup_on_exit))
+  if (!caml_startup_aux(0)) {
```
What is the point of this `pooling` parameter if it is always determined by the `caml_cleanup_on_exit` variable, either in the caller or the callee? I have the impression that we could either (1) get rid of it completely, or (2) never mutate it from within a function but always pass `caml_cleanup_on_exit` when entering the first function in a chain that is parametrized on `pooling`. Both would result in cleaner code than the odd mix we have today, and in particular I don't understand this `0` value here.

gadmm replied:
My answer was going to be that it is actually the logical OR of caml_cleanup_on_exit (obtained from parsing OCAMLRUNPARAM) and the pooling parameter (non-zero in other callers of caml_startup_aux).

Looking more closely, your objection might be due to the fact that this does not make sense anyway because caml_cleanup_on_exit needs to be updated to correctly call caml_shutdown (bug that was there before): instead of 
```c
  if (caml_cleanup_on_exit)
    pooling = 1;
  if (pooling)
    caml_stat_create_pool();
```
it should be
```c
  if (pooled)
    caml_cleanup_on_exit = 1;
  if (caml_cleanup_on_exit)
    caml_stat_create_pool();
```
Currently, caml_startup_pooled does not release memory on shutdown (and it never did) when its was its sole purpose.

Anyway, grepping the opam repo shows that nobody uses caml_startup_pooled. (I also failed to find evidence that anyone uses the OCAMLRUNPARAM `c` but this is harder to make sure of.)

gasche replied:
> Anyway, grepping the opam repo shows that nobody uses caml_startup_pooled.

This may be a bit misleading because we expect these settings to be decided by end-user programs, not libraries, and it much more common to distribute support libraries on opam than programs. The propose use-case to introduce pooled startup (#71) was to support dynamic loading and unloading of OCaml plugins written for a "properietary Delphi software"; I would not expect to find those on opam!

gadmm replied:
This is true. I know the caveats of grepping opam, and I had the usual precautions in my mind when I wrote that, but it is important to remind those caveats to not give false impressions. Thanks for the info as well.

murmour replied:
> Looking more closely, your objection might be due to the fact that this does not make sense anyway because caml_cleanup_on_exit **needs to** be updated to correctly call caml_shutdown

It doesn't need to.

Actually, `caml_cleanup_on_exit` (which is a state of a flag set via OCAMLRUNPARAM) is a superset of the pooling mode. Not only it enables pooling, but also enables shutdown on `exit` (which isn't desirable in all situations):

```c
CAMLprim value caml_sys_exit(value retcode_v)
{
  ...
  if (caml_cleanup_on_exit)
    caml_shutdown();
  ...
}
```


> (bug that was there before)
> [...]  
> Currently, caml_startup_pooled does not release memory on shutdown (and it never did) when its was its sole purpose.

This doesn't seem true. I haven't tested this code since 4.06 (which I use), but looking at it right now (before your patch), it seems quite obvious that the pool gets initialized in `caml_startup_pooled` (via `caml_startup_aux(1)`), and that `caml_shutdown` frees it thereafter (if you don't forget calling it). 

Please, double-check this in `trunk`. Perhaps you broke something with your patch.

gadmm replied:
Indeed, your explanations make sense to me. I saw the lines above, did not have the right expected behaviour in mind, and misunderstood what they did. I will remove the corresponding change and if there is something to clarify in the documentation then I will write something about it (it could be entirely my own misreading).

gasche reviewed (commented):
There are various things I don't understand with this change. I guess the main point is that there are several changes and it is not clear what does what.

The description announces that the PR fixes a crash, but how do we observe the crash? (Is it something that we should consider adding in our testsuite as a regression test?)

More importantly: before `caml_startup_aux` had a very simple role, it would count the number of times it was called to cutoff initialization, and also call `caml_stat_create_pool` which is related to the startup/shutdown logic. With your PR, a lot of the "init" logic is moved inside this function (why? this is not explained; is it to fix a bug or for code factorization?), but some of the logic remains in `startup_byt` and `startup_nat`. Looking at the diff there is no clear reason for what is moved where.

My guess would be that what stays in `startup_nat` is exactly, in your design, the initialization functions that need to be implemented differently in the bytecode and native runtime. Here is maybe a naming scheme that could make this simpler to follow/understand:
- `caml_starting_aux` is left as it was before, and possibly renamed into `caml_on_first_startup`
- the new stuff you moved in it moves to a different `caml_init_aux` function
- the remaining `init_*` bits in the byt/nat `caml_startup_*` functions are moved to `caml_init_{byt,nat}` functions, that start by calling `caml_init_aux` (so as a function they are in charge of all initialization)

gasche commented:
Ah, I realize now that some of my questions are answered by reviewing commit-by-commit, I had only read the whole diff. So if I understand correctly, the main fix is to ensure that `caml_init_domain` is not called several times. I think that my comments generally still apply.

gadmm commented:
Mea culpa, I forgot to describe the main bug and fix: the second commit moves functions that use caml_stat_alloc after the initialisation of the pool. Initialising the pool after something has been allocated will prevent the latter from being deallocated on shutdown, and will cause UB (most likely a crash) when attempting to free it manually. The logic of startup is a bit convoluted and recent evolutions forgot to take this logic into account (also it is hard to see how this can be caught by testing, but I think adding comments in the source will avoid these issues in the future).

gasche commented:
> also it is hard to see how this can be caught by testing

We could expose `caml_stat_alloc_unpooled` and `caml_stat_alloc_outside_pool`. The `unpooled` version would set a global flag that unpooled allocations happened (this is bad if the runtime intends to be pooled), possibly only in the debug runtime. The `outside_pool` version is for allocations we explicitly want to happen outside the pool even in pooled mode (for example #10250, #10266). The `pool == NULL` fallback of `caml_stat_alloc` would call `caml_stat_alloc_unpooled`.

Then a test could a C program starting the pooled runtime, interpreting a toy bytecode program, shutting down the runtime and cleaning the pool, then checking that the `unpooled` flag is not set.

gadmm commented:
I started implementing the idea of a flag but then realised that for people who call caml_stat_alloc*_noexc functions without the runtime lock, this flag would need to be synchronised. The documentation of some _noexc functions explicitly mention that the runtime lock is not needed. The pooled mode was not made the default because of backwards compatibility concerns (with lots of positives found in the original PR), but in fact the new requirements were not added to the documentation (again: in pooled mode you *need* the runtime lock to call the _noexc functions) and are not enforced in any way with default OCaml settings, so library writers are unlikely to have started to follow the new discipline. These requirements are not even necessarily easy to follow, think about needing to acquire the runtime lock (and thus knowing whether you already have it) if you need to free a block given to foreign code during clean-up after an error condition (when the documentation of caml_stat_free does not even mention needing the runtime lock).

It seems to me that a better course of action would be to:
- introduce `caml_stat_alloc_pooled*` functions (they would fail if the pool is not initialised),
- use pooled functions everwhere in the runtime (except at the few places where one needs explicit management),
- remove the optional pooled mode which is no longer necessary to prevent leaks when shutting down the runtime.

The `caml_stat_alloc*` functions would recover their original semantics compatible with the current documentation, and libraries could start using the pooled allocations explicitly if they wanted. And the memory allocated by the runtime will be freed for everybody.

I want to stress that to take advantage of the pooled allocations the program or library must be explicitly designed for it. Indeed, for a library to correctly release memory allocated with caml_stat_alloc*, it must still call caml_stat_free explicitly because pooling is not the default behaviour. In this scenario, when activating the pooled mode, either the pooled mode has no useful effect (if the deletions are all done before shutdown and with runtime lock held), or it always causes UB (otherwise). So it does not make sense to have a switch to choose whether allocations are pooled or not.

Anyway, these comments go fairly beyond the purpose of this PR. For the current PR I prefer to stop at its current state I think.

gasche commented:
Sounds reasonable, thanks. Do you have an opinion on the refactoring of the `init` functions proposed in [my review comment](https://github.com/ocaml/ocaml/pull/10304#pullrequestreview-616984669)?

gadmm commented:
Yes, I do not see what is wrong about moving stuff to startup_aux. Even if it had a specific purpose I do not see the issue with repurposing it.

gasche commented:
The function moves from "one purpose" to "two purposes" (what it was doing now + some of the runtime initialization logic), so it is getting strictly more complex, which is bad.

gasche commented:
cc @murmour: do you still care about pooled allocations? What do you think of @gadmm's different API proposal? (Would you want to implement it?)

xavierleroy commented:
> (I also failed to find evidence that anyone uses the OCAMLRUNPARAM c but this is harder to make sure of.)

For the record: the "c" parameter is tested as part of the CI at Inria, in the "other configs" script:
https://github.com/ocaml/ocaml/blob/414bdec9ae387129b8102cc6bf3c0b6ae173eeb9/tools/ci/inria/other-configs/script#L46

and also in the "sanitizers" script, in conjunction with a leak detector:
https://github.com/ocaml/ocaml/blob/414bdec9ae387129b8102cc6bf3c0b6ae173eeb9/tools/ci/inria/sanitizers/script#L88-L94

So I'm reasonably confident that the "c" mode works and is maintained.  The "pooled" API is another matter, though.

murmour commented:
> cc @murmour: do you still care about pooled allocations? What do you think of @gadmm's different API proposal? (Would you want to implement it?)

Hello, Gabriel, Guillaume, and Monsieur Leroy!

I still use the pooled (unloadable) runtime in 4.06, and totally care about it.

This PR is quite interesting, and I will do a proper review a bit later (in a few days).

gadmm commented:
This got me puzzled as to why it did not catch that `Caml_state` was allocated before pool initialisation. Digging deeper, the [tool](https://github.com/google/sanitizers/wiki/AddressSanitizerLeakSanitizerDesignDocument) considers global variables as live at the end of the program, so it did not consider `Caml_state` as leaking. I did not know, but it does not catch similar leaks (including anything reachable from globals, thread-local variables, etc.)

gadmm committed:
More factoring and simplifications

The calls to caml_init_atom_table can be factored.

Then, make static what can be.

gadmm commented:
I have pushed some simplifications (make static what does not need to be extern, etc.). I have also found a hidden function that could be factored (init_atom_table). Its place inside init_static in nat was nice and appropriate but I think it is even better to factor the call between nat and byt and move it to the common file.

As discussed with @gasche I do not see how to easily improve the organisation further but if you want to show me what you have in mind I will be happy to consider it as a patch.

gasche commented:
See https://github.com/gadmm/ocaml/pull/1 for a modest proposal.

gadmm committed:
Clarify documentation for caml_shutdown

Fix #10179 (wrong example of thing that is released on shutdown), and
clarify the difference between startup_pooled and the
"cleanup_on_exit" function (at least to me).

Fix typo.

gadmm commented:
I added @gasche's suggestion and take advantage of this open PR to introduce a minor clarification to the documentation (closes #10179). This is hopefully the last addition to this PR, further reviews welcome!

murmour commented on runtime/startup_aux.c:
```diff
@@ -168,14 +168,14 @@ int caml_startup_aux(int pooling)
-  if (caml_cleanup_on_exit)
-    pooling = 1;
+  if (pooling)
+    caml_cleanup_on_exit = 1;
```
I don't see the meaning in this change.

Previously, if the user set `OCAMLRUNPARAM=c`, then `caml_cleanup_on_exit` became 1, which enabled pooling.

Now, if pooling is enabled (via a call to `caml_startup_pooled`), it sets `caml_cleanup_on_exit` to 1, which later forces `caml_shutdown` on `exit`. But the user didn't ask for this! Leak detection mode (`caml_cleanup_on_exit`) isn't the same as ability to unload the runtime manually.

gadmm replied:
I have removed this change, which stemmed from my misunderstanding. I also add a clarification of the difference between the two features in the documentation if you accept, to explain what you just wrote.

murmour commented on manual/src/cmds/intf-c.etex:
```diff
@@ -1753,9 +1753,8 @@ gracefully, which equals the following:
 \begin{itemize}
```
While you're here, please fix "4.05" to "4.06" on line 1751 (GitHub doesn't let me comment there).

gadmm replied:
Done.

murmour reviewed (changes requested):
General comment: this PR lacks structure, which made it difficult to review. It is good that the changes are split into commits, but these commits lack proper messages (with explanations), and aligning these commits with the PR comments required effort.

`Changes` entry could be more detailed. Perhaps it should be split into 2-3 entries.

---

> In addition, the last commit prevents reparsing OCAMLRUNPARAM on subsequent calls to caml_startup*. In rare conditions, e.g. when using setenv/putenv to change the value of caml_use_huge_pages after OCaml has already started, this can lead to UB (a segfault most probably). This is unlikely to happen, but it shows that it does not make sense to reparse OCAMLRUNPARAM.
> [...]
> Mea culpa, I forgot to describe the main bug and fix: the second commit moves functions that use caml_stat_alloc after the initialisation of the pool. Initialising the pool after something has been allocated will prevent the latter from being deallocated on shutdown, and will cause UB (most likely a crash) when attempting to free it manually. The logic of startup is a bit convoluted and recent evolutions forgot to take this logic into account (also it is hard to see how this can be caught by testing, but I think adding comments in the source will avoid these issues in the future).

Correct. Both of these problems (an unlikely UB and a few minor leaks) were introduced by me in cef4a941c6e7b5ff423f4e9243bb66f42745e344, while adding `caml_cleanup_on_exit`.

---

> See gadmm#1 for a modest proposal.

What strikes me wrong with this approach is that `caml_{startup,shutdown}_needed` are named like predicates, but actually increment/decrement global counters on calls, which is completely unobvious.

How about just renaming `caml_stratup_aux` to `caml_startup_common` and documenting it properly?

---

> I [...] take advantage of this open PR to introduce a minor clarification to the documentation (closes #10179).

This change is correct and fixes a mistake that I added in af5899fa8b0ae18a80e2e50d512c5d4cce4c5f28.

gadmm commented:
I will come back to this PR in a moment. In the meanwhile, I'd like to know what you think @gasche :

>> See gadmm#1 for a modest proposal.
>
> What strikes me wrong with this approach is that caml_{startup,shutdown}_needed are named like predicates, but actually increment/decrement global counters on calls, which is completely unobvious.
>
> How about just renaming caml_stratup_aux to caml_startup_common and documenting it properly?

I tend to agree with @murmour, for this reason and because in the end the code to call the two functions ends up duplicated between the three call sites. Would this alternative be fine with you?

gadmm commented:
@murmour :

> Both of these problems (an unlikely UB and a few minor leaks) were introduced by me in cef4a94, while adding caml_cleanup_on_exit.

To make sure I did not miss anything, the bug with Caml_state and the event buffer being allocated before pool initialization is more recent, right?

gasche commented:
>  Would this alternative be fine with you?

I think; I timed out on this problem and I am okay with a code proposal that you and @murmour would both like.

murmour commented:
> @murmour :
> 
> > Both of these problems (an unlikely UB and a few minor leaks) were introduced by me in [cef4a94](https://github.com/ocaml/ocaml/commit/cef4a941c6e7b5ff423f4e9243bb66f42745e344), while adding caml_cleanup_on_exit.
> 
> To make sure I did not miss anything, the bug with Caml_state and the event buffer being allocated before pool initialization is more recent, right?

Right. It was introduced in eac4860ad387156e7bf1ebee17c86ac9eb52af4c (#8713, 4.10.0).

murmour commented:
@gadmm, regarding your ideas on `caml_stat_*` API: it's important stuff, and I'll give thorough feedback on it a bit later (running short of time right now). It is out of scope of this PR anyway.

gadmm committed:
Better name for caml_startup_aux

Also the previous caml_startup_common should be static; rename it.

gadmm commented:
Here is a rebased version. Comments have been added in the logs, otherwise only the last three commits have changed.

Thanks for the review. I took your comments into account.

You are right that I could have made your task easier by including detailed commit logs and mentioning that it is best reviewed commit-per-commit. I will be more careful next time.

I do not see how to improve the Changes entry. I think all the user-visible changes have been described.

gadmm commented:
The failing test appears unrelated.

gadmm commented:
Happy belated birthday, PR#10304!

I am closing this PR since I no longer have time not interest in this problem in light of the absence of review and the work needed to rebase it on top of multicore (and duplication of work to fix the issues in both OCaml 4 and 5).

I have tried to use the pooled mode for our library, and as explained in https://github.com/ocaml/ocaml/pull/10304#issuecomment-803631573, this mode needs a rework before it can be useful, so we are going to do without it. For now I simply recommend to stay clear from the "`c`" runparam.

But anyone who cares about these issues should feel free to adopt it and/or open an issue about it. I am not deleting the branch so the code will remain available for now.
